### ISSUE TITLE
Layers for Text Classification

### ISSUE THREAD

**Issue by bokstaller** [flairNLP/flair#2567]
Hi,

when I am training the Flair text classifier with an embedding model (e.g. RoBERTa) how is the classification model generated? As far as I understood is, that the embedding model outputs the embedding for each token which goes into the classifier that predicts the class. But how can I find out, what the architecture of the classification model looks like because I need to modify it in such a way to use the triplet loss.

Thanks in advance,

Jonas.

**Comment by whoisjones** [comment]
hi @bokstaller,

regarding your architecture question, this is from our TextClassifier constructor:
```
self.document_embeddings: flair.embeddings.DocumentEmbeddings = document_embeddings

self._label_type = label_type

self.decoder = nn.Linear(self.document_embeddings.embedding_length, len(self.label_dictionary))
```
You see we map from the embedding dimension to the tag space.

Regarding your second question, TextClassifier inherits from DefaultClassifer that uses standard loss functions:
```
if self.multi_label:
    self.loss_function: _Loss = torch.nn.BCEWithLogitsLoss(weight=self.loss_weights)
else:
    self.loss_function = torch.nn.CrossEntropyLoss(weight=self.loss_weights)
```

In order to get TripletLoss, you can create a new classifier class, that inherits from DefaultClassifier and overwrite DefaultClassifier's `_calculate_loss()` method (this is called during training and calls respective loss_function attribute of the model). Note that you need to overwrite for every model the `forward_pass()` method.

**Comment by bokstaller** [comment]
Hi @whoisjones,

thanks for your fast reply. So if I understand correctly, basically I just need to copy the ### _calculate_loss ### method to the ### text_classification_model ### class and instead of ### return self.loss_function(scores, labels), len(labels) ### I could use e.g. ###torch.nn.TripletMarginLoss###.

Best,

Jonas

**Comment by whoisjones** [comment]
hi @bokstaller,
flair represents different models through inheritance, i.e. all standard classification tasks inherit from `DefaultClassifier`, because they share various similar functions such as `predict()` or `evaluate()`. This way one is required to overwrite only the `forward_pass()` method of a model to make it work. For you solution to work, I suggest to create a new model class that inherits from `DefaultClassifier`, you would need to overwrite `forward_loss()` (called by trainer), `forward_pass()` (computes and returns scores + gold labels) and `_calculate_loss()` (uses your TripletLoss).

something like (haven't tested it):
```
import flair
from flair.nn import DefaultClassifier
from flair.embeddings import DocumentEmbeddings

import torch


class TripletLossClassifier(DefaultClassifier):

    def __init__(self, document_embeddings: flair.embeddings.DocumentEmbeddings, label_type: str, **classifierargs):
        super().__init__(document_embeddings, label_type, **classifierargs)
        self.loss_function = torch.nn.TripletMarginLoss(margin=1.0, p=2)
        self.loss_function = torch.nn.TripletMarginLoss(margin=1, p=2)

    def forward_loss(self, sentences: Union[List[DT], DT]):
        scores, labels = self.forward_pass(sentences)
        return self._calculate_loss(scores, labels)

    def forward_pass(self, sentences):
        #TODO implement forward method
        return scores
    
    def _calculate_loss(self, scores, labels):
        return self.loss_function(scores, labels)
```

**Comment by bokstaller** [comment]
Great, thanks for the code snipping, that clears up a lot for me! But do I need to also implement my own forward pass, or can I just use the standard one from the text classifier, because as far as I understood it, I don't need to change anything in the forward pass?
